Standardizing arguments passed to callback functions

Bernstern opened the pull request:
Implementation for standardizing callback arguments passed to callback functions to make use easier across platforms (resolves #301).

This implements the feature for all platforms and include doc updates and an example.

hbldh commented:
I like the idea, but I have not got the time to review this right now; it is a lot of changes...

@Bernstern Could you fix the conflicts and squash it down to one commit and I will look this over next week?

dlech commented on bleak/backends/dotnet/scanner.py:
```diff
@@ -83,8 +85,50 @@ def AdvertisementWatcher_Received(self, sender, e):
+                callback_data["name"] = e.Advertisement.LocalName or None
+                callback_data["data_channel"] = None
+                callback_data["manufacturer_data"] = data
+                callback_data["service_data"] = None
```
I think we should be able to get this with `e.Advertisement.GetSectionsByType(0x16)` plus some additional processing. And 0x20 and 0x21 are probably needed too to be complete.

dlech reviewed (changes requested):
Thanks for taking time to work on this. I've made a bunch of comments, but looking at this from a higher level, we are already parsing advertising data in 2 places (implementations of `BLEDevice` and `BaseBleakScanner` abstract classes) and this adds a 3rd. Reusing this existing code would fix many of the bugs I pointed out in this PR.

I think we could create an `AdvertisementData` class based on the dictionary given in this PR and utility functions in each backend to convert the platform-specific data into the common `AdvertisementData` type. Then this common type could be passed around instead of the platform-specific data.

dlech commented on bleak/backends/dotnet/scanner.py:
```diff
@@ -1,10 +1,12 @@
+import inspect
```
Doesn't look like this is used anywhere.

Bernstern commented:
I agree, having it as a separate class will be really handy, I'll implement those changes (and remove that debug print :smiley: ) over the next few days.

Bernstern commented:
I implemented the requested changes to use an AdvertisementData class and updated the documentation and the example I wrote. I didn't quite follow @dlech's suggestion to include the other advertisement sections for the dotnet backend so I'm open to suggestions on how to implement that :smiley:.

Bernstern commented on bleak/backends/corebluetooth/scanner.py:
```diff
@@ -46,10 +45,93 @@ async def start(self):
+            callback_data = get_reference_callback_format()
+
+            # Get the raw data from the advertisement to be parse
+            service_data_dict_raw = a.get("kCBAdvDataServiceData", {})
```
Updated the service data to use the suggested dictionary comprehension and conversion closer.

dlech commented on bleak/backends/bluezdbus/scanner.py:
```diff
@@ -304,4 +303,38 @@ def parse_msg(self, message):
         if self._callback is not None:
-            self._callback(message)
+            # First ensure that the message body is a list and has the dict where we can find the information we need
+            if isinstance(message.body, list) and len(message.body) > 2:
```
When would this be false? Are there other kinds of messages that are received?

If it is not expected to ever be false, we should probably remove the check. Otherwise, it would be nice to invert the check so that the error is handled in the indented if statement and the normal code flow is not indented.

Bernstern replied:
I added the check to be safe with accessing the list in the next line, I haven't seen a case yet where it didn't pass so it should be safe to remove.

dlech commented on bleak/backends/bluezdbus/scanner.py:
```diff
@@ -304,4 +303,38 @@ def parse_msg(self, message):
+                discovery_data = message.body[1]
+
+                # Get all the information wanted to pack in the advertisement data
+                _address = message.path[message.path.rfind("dev_") + 4 :].replace(
```
Why not use 

    _address = discovery_data["Address"]

instead of scraping the address from the D-Bus path?

dlech commented on bleak/backends/bluezdbus/scanner.py:
```diff
@@ -304,4 +303,38 @@ def parse_msg(self, message):
+                    "_", ":"
+                )
+
+                _local_name = discovery_data.get("Name", None)
```
This could be 

    _local_name = discovery_data.get("Name", "Unknown")

instead of the special case for  `_local_name`  below

dlech commented on bleak/backends/corebluetooth/scanner.py:
```diff
@@ -40,11 +39,65 @@ def __init__(self, **kwargs):
     async def start(self):
         self._identifiers = {}
 
+        def safe_list_get(l: list, idx: int, default):
```
It looks like this method is unused and can be removed.

dlech commented on bleak/backends/corebluetooth/scanner.py:
```diff
@@ -40,11 +39,65 @@ def __init__(self, **kwargs):
+            except IndexError:
+                return default
+
+        def callback_dict_breakdown(data) -> Union[None, dict]:
```
This is also not used and can be removed.

dlech commented on bleak/backends/corebluetooth/scanner.py:
```diff
@@ -40,11 +39,65 @@ def __init__(self, **kwargs):
+                    manufacturer_binary_data[0:2], byteorder="little"
+                )
+                manufacturer_value = bytes(manufacturer_binary_data[2:])
+                manufacturer_data = {manufacturer_id: manufacturer_value}
```
This could modify the existing dictionary instead of creating a new one

    manufacturer_data[manufacturer_id] = manufacturer_value

dlech commented on bleak/backends/scanner.py:
```diff
@@ -111,3 +111,34 @@ async def _find_device_by_address(
+class AdvertisementData:
+    """
+    Wrapper around the advertisement data that each platform returns upon discovery
+    """
```
Would be nice to list the allowable kwags in the docstring.

Bernstern replied:
Updated to include required and keyword args

dlech commented on docs/scanning.rst:
```diff
@@ -96,6 +96,10 @@ In the manual mode, it is possible to add an own callback that you want to call
 which differ widely between OS backend implementations, so the instructions merit careful reading.
 
+Note that the when using :code:`BleakScanner.register_detection_callback(callback)`, your callback
+function will be passed a AdvertisementData class whose structure can be found in the base scanner file.
```
an :class:`AdvertismentData`

dlech commented on docs/scanning.rst:
```diff
@@ -96,6 +96,10 @@ In the manual mode, it is possible to add an own callback that you want to call
 
+Note that the when using :code:`BleakScanner.register_detection_callback(callback)`, your callback
+function will be passed a AdvertisementData class whose structure can be found in the base scanner file.
+`See detection_callback.py` example. Note that platform specific advertisement data can be found in the platform_data field.
```
``platform_data``

dlech commented on examples/detection_callback.py:
```diff
@@ -0,0 +1,34 @@
+
+    while True:
+        await scanner.start()
+        await asyncio.sleep(1)
```
one second seems a bit short for a demo

dlech reviewed (changes requested):
This looks good. Just a few more questions and suggestions.

We can look at reusing this to replace the duplicated code later.

Bernstern commented:
Let me know if there are any other changes you would like to see.

dlech commented:
I was hoping to have to have some time to test this before merging, but I haven't had time yet.

dlech commented on examples/detection_callback.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def simple_callback(callback_data: AdvertisementData):
+    print(callback_data)
```
it would be nice to print all data here (e.g. the suggested `__repr__` method above)

dlech commented on bleak/backends/dotnet/scanner.py:
```diff
@@ -89,9 +90,23 @@ def _received_handler(self, sender, event_args):
         if self._callback is not None:
-            self._callback(sender, event_args)
+            # Get a "BLEDevice" from parse_event args
+            temporary_device = self.parse_eventargs(event_args)
```
nitpick: `temporary_device` seems a bit verbose, `device` should do just fine

dlech commented on bleak/backends/dotnet/scanner.py:
```diff
@@ -89,9 +90,23 @@ def _received_handler(self, sender, event_args):
+                local_name=temporary_device.name or "Unknown",
+                rssi=temporary_device.rssi,
+                manufacturer_data=temporary_device.metadata["manufacturer_data"],
+                service_data=event_args.Advertisement.GetSectionsByType(0x16),
```
The service data needs to be parsed to put it in the same format as other platforms. Here is what I came up with:

```python
            # Decode service data
            service_data = {}
            # 0x16 is service data with 16-bit UUID
            for section in event_args.Advertisement.GetSectionsByType(0x16):
                with BleakDataReader(section.Data) as reader:
                    data = reader.read()
                    service_data[
                        f"0000{data[1]:02x}{data[0]:02x}-0000-1000-8000-00805f9b34fb"
                    ] = data[2:]
            # 0x20 is service data with 32-bit UUID
            for section in event_args.Advertisement.GetSectionsByType(0x20):
                with BleakDataReader(section.Data) as reader:
                    data = reader.read()
                    service_data[
                        f"{data[3]:02x}{data[2]:02x}{data[1]:02x}{data[0]:02x}-0000-1000-8000-00805f9b34fb"
                    ] = data[4:]
            # 0x21 is service data with 128-bit UUID
            for section in event_args.Advertisement.GetSectionsByType(0x21):
                with BleakDataReader(section.Data) as reader:
                    data = reader.read()
                    service_data[str(UUID(bytes_le=data[:16]))] = data[16:]
```

Bernstern replied:
Thank you for your help with this!

dlech commented on bleak/backends/dotnet/scanner.py:
```diff
@@ -89,9 +90,23 @@ def _received_handler(self, sender, event_args):
+            self._callback(advertisement_data)
 
-    def _stopped_handler(self, sender, event_args):
+    def _stopped_handler(self, sender, e):
```
`_` would be better than `e` if the argument is not used

dlech commented on bleak/backends/bluezdbus/scanner.py:
```diff
@@ -304,4 +303,34 @@ def parse_msg(self, message):
+            # Get all the information wanted to pack in the advertisement data
+            _address = discovery_data["Address"]
+
+            _local_name = discovery_data.get("Name", "Unknown")
```
Also should be using `"Alias"` instead of `"Name"`

dlech replied:
There is also a new experimental "AdvertisementData" property in BlueZ that we could use to get the local name instead of assuming the "Name" property came from the advertising data (which it may or may not have).

https://github.com/bluez/bluez/blob/master/doc/device-api.txt#L260

We don't have to implement this now, but we could add a TODO comment to remind us about it.

dlech reviewed (changes requested):
Finally had some time to get back to this. It is looking really good.

Although I did find some problems on Windows and Linux - see inline comments.

FWIW, you can find my working branch at https://github.com/dlech/bleak/commits/pr/Bernstern/334 which contains the suggested fixes and some other unrelated fixes

dlech commented:
@hbldh do you have an opinion on where the `AdvertisementData` class should live? Currently it is at `bleak.backends.scanner.AdvertisementData`.

Bernstern commented on bleak/backends/bluezdbus/scanner.py:
```diff
@@ -304,4 +303,34 @@ def parse_msg(self, message):
 
         if self._callback is not None:
-            self._callback(message)
+            discovery_data = message.body[1]
```
Sgtm! I saw similar behavior where a bunch of callback messages were just the RSSI updates or similar, this is a really clean solution

Bernstern commented:
@dlech Thank you very much for your suggestions and your working branch made it very apply your requested changes and test them locally, much appreciated! 

I added all of the changes on the working branch and merged develop back into it since there were merge conflicts but they should be resolved.

dlech commented:
Awesome. Let's call this good enough and get it merged.
